### PR TITLE
Implemented proper support for TAI and UTC.

### PR DESCRIPTION
**Pull request by davidv1992**

### CODE REVIEW

**Comment by folkertdev** [comment]
I did a trivial rebase (I think you cherry-picked some commits and github is not smart enough to realize that there are no conflicts)

File: statime-linux/src/clock/mod.rs [hunk 1]
```diff
@@ -80,7 +74,15 @@ impl Clock for LinuxClock {
     fn set_frequency(&mut self, freq: f64) -> Result<Time, Self::Error> {
         use clock_steering::Clock;
         log::trace!("Setting clock frequency to {:e}ppm", freq);
-        let timestamp = self.clock.set_frequency(freq)?;
+        let timestamp = if self.is_tai {
+            // Clock tai can't directly adjust frequency, so drive this through
+            // clock_realtime and adjust the received timestamp
+            let mut ts = UnixClock::CLOCK_REALTIME.set_frequency(freq)?;
+            ts.seconds += UnixClock::CLOCK_REALTIME.get_tai()? as libc::time_t;
```
Comment (folkertdev) [inline statime-linux/src/clock/mod.rs:81]: so, this function does not really `get_tai`, but `get_tai_offset` right? (in other words, the number of leap seconds that has been applied). I think that function name is confusing

File: statime-linux/src/clock/mod.rs [hunk 2]
```diff
@@ -7,55 +7,49 @@ use statime::{Clock, Duration, Time, TimePropertiesDS};
...
+        self.clock.system_offset().map(|(mut t1, t2, mut t3)| {
+            let tai_offset = UnixClock::CLOCK_REALTIME.get_tai().unwrap();
+            t1.seconds += tai_offset as libc::time_t;
+            t3.seconds += tai_offset as libc::time_t;
+            (
+                clock_timestamp_to_time(t1),
+                clock_timestamp_to_time(t2),
+                clock_timestamp_to_time(t3),
+            )
+        })
```
Comment (folkertdev) [inline statime-linux/src/clock/mod.rs:39]: this function is confusing. There is a lot of subtle `tai` stuff going on, and I believe you but do not fully understand why this is happening?
```suggestion
    /// Some description of what we return here
    pub fn system_offset(&self) -> Result<(Time, Time, Time), clock_steering::unix::Error> {
        // what do we get back?
        let (mut t1, t2, mut t3) = self.clock.system_offset()?;
        let tai_offset = UnixClock::CLOCK_REALTIME.get_tai()?; // function should be `get_tai_offset`

        t1.seconds += tai_offset as libc::time_t;
        t3.seconds += tai_offset as libc::time_t;

        Ok((
            clock_timestamp_to_time(t1),
            clock_timestamp_to_time(t2),
            clock_timestamp_to_time(t3),
        ))
```

File: statime-linux/src/clock/mod.rs (hunk 1 above)
Comment (davidv1992) [reply statime-linux/src/clock/mod.rs:81]: Fair, though that needs changing then in the clock crate. Suggest we keep it as is for now.

File: statime-linux/src/clock/mod.rs (hunk 2 above)
Comment (davidv1992) [reply statime-linux/src/clock/mod.rs:39]: Added explanation. for the get_tai stuff see above

**Review by folkertdev — APPROVED** [review]

**Pull request merged**